compiler/AST/iterator.cpp

```diff
@@ -68,6 +69,267 @@ IteratorInfo::IteratorInfo() :
+  It is accessible using getIteratorGroup(Iterator).
+
+  Given a group IG, getIteratorGroup(IT2)==IG for each iterator IT2
+  that IG points to: IG.serial, IG.standalone (when non-NULL), etc.
+
+  IG.serial is always non-NULL.
+  
+* An iterator IT2 (any flavor) is pointed to from a group IG
+  if and only if  getIteratorGroup(IT2) == IG.
+
+SERIAL / STANDALONE / LEADER
+
+* If IT is a serial iterator, there is an iterator group IG for it
+  such that:
+   - IG.serial points to IT.
+   - IG.standalone points to the corresponding standalone iterator
+     if it is available, NULL otherwise.
+   - IG.leader - ditto.
+   - IG.follower is currently unused.
+
+* If a serial IT has no corresponding standalone or leader,
+  there is still a group IG for IT, such that:
+   - IG.seral == IT
+   - IG.standalone == IG.leader == NULL.
+  This can be used to check availability of standalone/leader.
+
+* If IT is a standalone or leader iterator, there may or may not be
+  a group for it. We are interested in a group only when it maps
+  from a serial iterator to its parallel counterparts. So when
+  a CallExpr invokes a parallel iterator directly, iterator groups
+  are not updated.
+
+SERIAL --> PARALLEL
+
+* Each parallel iterators is attempted to be resolved by resolving
+  a "representative call". It is created as follows:
+   - It seeks a function of the same name as the serial iterator's.
+   - The actual arguments are the serial iterator's formals, plus 'tag'.
+
+* That way there is a 1:1 relationship between the serial iterator
+  its standalone counterpart; ditto leader. Because the choice
+  of the parallel iterator is affected only by the serial iterator.
+  This choice is not affected by the "original call" i.e. one that
+  led to resolving the serial iterator.
+
+* The "representative call" is placed next to the "original call".
+  To strengthen confidence in this 1:1 relationship, it seems like
+  the representative call should go next to the definition
+  of the serial iterator. However, that would cause visibility issues.
+
+LOGISTICS  
+
+* Only the availability of the standalone and leader iterators
+  is detected. Their bodies are not resolved, to avoid encountering
+  (and generating) potential compile-time errors when those iterators
+  are not actually used in the program.
+
+* The absence of a parallel iterator is not an error. An error may
+  be issued later when+if this iterator is required for a parallel
+  computation.
+
+* The follower iterator is not sought at this time.  to seek it we would
+  need to know what the leader iterator yields. For that we may have to
+  resolve the **body** of the leader. Which we are not doing here.
+
+*/
+
+std::map<FnSymbol*,IteratorGroup*> iteratorGroups;
+
+IteratorGroup::IteratorGroup() :
+  serial(NULL), standalone(NULL), leader(NULL), follower(NULL) {}
+
+static bool isGroupable(FnSymbol* it) {
+  return it->hasFlag(FLAG_ITERATOR_FN) ||
+         it->hasFlag(FLAG_FN_RETURNS_ITERATOR);
```

Comment:
`isGroupable` seems like a poor choice of name for this function. I think `isIteratorLike`, or something along those lines, would be more reasonable.

Reply:
+1 on the new name